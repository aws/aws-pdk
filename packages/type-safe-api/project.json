{
  "name": "@aws/type-safe-api",
  "root": "packages/type-safe-api",
  "targets": {
    "build": {
      "inputs": [
        "default",
        "^default",
        "!{projectRoot}/.jsii",
        "!{projectRoot}/lib/**/*",
        "!{projectRoot}/dist/**/*",
        "!{projectRoot}/coverage/**/*",
        "!{projectRoot}/test-reports/**/*",
        "!{projectRoot}/docs/api/**/*"
      ],
      "outputs": [
        "{projectRoot}/.jsii",
        "{projectRoot}/lib",
        "{projectRoot}/dist",
        "{projectRoot}/coverage",
        "{projectRoot}/test-reports",
        "{projectRoot}/docs/api"
      ],
      "dependsOn": [
        "^build"
      ],
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen build",
        "cwd": "packages/type-safe-api"
      }
    },
    "default": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen default",
        "cwd": "packages/type-safe-api"
      }
    },
    "pre-compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen pre-compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "post-compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen post-compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "test": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen test",
        "cwd": "packages/type-safe-api"
      }
    },
    "package": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package",
        "cwd": "packages/type-safe-api"
      }
    },
    "test:watch": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen test:watch",
        "cwd": "packages/type-safe-api"
      }
    },
    "watch": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen watch",
        "cwd": "packages/type-safe-api"
      }
    },
    "eslint": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen eslint",
        "cwd": "packages/type-safe-api"
      }
    },
    "compat": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen compat",
        "cwd": "packages/type-safe-api"
      }
    },
    "package-all": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package-all",
        "cwd": "packages/type-safe-api"
      }
    },
    "package:java": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package:java",
        "cwd": "packages/type-safe-api"
      }
    },
    "package:python": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package:python",
        "cwd": "packages/type-safe-api"
      }
    },
    "eslint-staged": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen eslint-staged",
        "cwd": "packages/type-safe-api"
      }
    },
    "jest": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen jest",
        "cwd": "packages/type-safe-api"
      }
    },
    "docgen": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen docgen",
        "cwd": "packages/type-safe-api"
      }
    }
  },
  "implicitDependencies": [
    "smithy-async-transformer"
  ],
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \"pnpm exec projen\"."
}
