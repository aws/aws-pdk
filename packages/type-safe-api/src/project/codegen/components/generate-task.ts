/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import { ProjectUtils } from "@aws/monorepo";
import { Component, Project, Task } from "projen";

/**
 * Component which manages a "generate" task for a project
 */
export class GenerateTask extends Component {
  /**
   * Retrieves an instance of GenerateTask if one is associated to the given project.
   *
   * @param project project instance.
   */
  static of(project: Project): Task | undefined {
    return (
      project.components.find((c) =>
        ProjectUtils.isNamedInstanceOf(c, GenerateTask)
      ) as GenerateTask | undefined
    )?.task;
  }

  /**
   * Retrieves an instance of GenerateTask if one is associated to the given project,
   * otherwise creates a GenerateTask instance for the project.
   *
   * @param project project instance.
   */
  static ensure(project: Project): Task {
    return GenerateTask.of(project) || new GenerateTask(project).task;
  }

  public readonly task: Task;

  constructor(project: Project) {
    super(project);

    this.task =
      project.tasks.tryFind("generate") ?? project.addTask("generate");
  }
}
