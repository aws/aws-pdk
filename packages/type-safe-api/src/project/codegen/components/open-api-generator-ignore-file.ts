/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import { IgnoreFile, Project } from "projen";

/**
 * Represents an .openapi-generator-ignore file
 * @see https://github.com/OpenAPITools/openapi-generator/blob/master/docs/customization.md#ignore-file-format
 */
export class OpenApiGeneratorIgnoreFile extends IgnoreFile {
  public static ALL_FILES_PATTERNS: string[] = ["/*", "**/*", "*"];

  constructor(project: Project) {
    super(project, ".openapi-generator-ignore");

    // We should always use the projen .gitignore since projen manages the build, and therefore the ignored build
    // artifacts
    this.addPatterns(".gitignore");
  }
}
