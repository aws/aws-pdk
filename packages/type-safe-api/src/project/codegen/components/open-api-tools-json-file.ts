/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import { ProjectUtils } from "@aws-prototyping-sdk/nx-monorepo";
import { JsonFile, Project } from "projen";
import { OpenApiGeneratorCliConfig } from "../../types";

/**
 * Represents an openapitools.json file
 * @see https://github.com/OpenAPITools/openapi-generator-cli#configuration
 */
export class OpenApiToolsJsonFile extends JsonFile {
  /**
   * Retrieves an instance of OpenApiToolsJsonFile if one is associated to the given project.
   *
   * @param project project instance.
   */
  static of(project: Project): OpenApiToolsJsonFile | undefined {
    return project.components.find((c) =>
      ProjectUtils.isNamedInstanceOf(c, OpenApiToolsJsonFile)
    ) as OpenApiToolsJsonFile | undefined;
  }

  /**
   * Retrieves an instance of OpenApiToolsJsonFile if one is associated to the given project,
   * otherwise creates a OpenApiToolsJsonFile instance for the project.
   *
   * @param project project instance.
   */
  static ensure(project: Project): OpenApiToolsJsonFile {
    return (
      OpenApiToolsJsonFile.of(project) || new OpenApiToolsJsonFile(project)
    );
  }

  /**
   * Configuration for OpenAPI Generator CLI
   * @private
   */
  private config: OpenApiGeneratorCliConfig = {
    version: "6.3.0",
    storageDir: "~/.open-api-generator-cli",
  };

  constructor(project: Project) {
    if (OpenApiToolsJsonFile.of(project)) {
      throw new Error(
        `Project ${project.name} already has associated OpenApiToolsJsonFile component.`
      );
    }

    super(project, "openapitools.json", {
      obj: {
        // Schema is located in node_modules when generator cli is installed in tmp dir
        $schema:
          "node_modules/@openapitools/openapi-generator-cli/config.schema.json",
        spaces: 2,
        "generator-cli": () => this.config,
      },
    });
  }

  /**
   * Update the OpenAPI Generator config. Shallow-merges with any existing default config
   */
  public addOpenApiGeneratorCliConfig(config?: OpenApiGeneratorCliConfig) {
    this.config = {
      ...this.config,
      ...config,
    };
  }
}
