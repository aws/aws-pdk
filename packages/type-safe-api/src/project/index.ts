/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
export * from "./model/type-safe-api-model-build";
export * from "./model/type-safe-api-async-model-build";
export * from "./model/smithy/smithy-project-definition";
export * from "./model/openapi/open-api-project-definition";
export * from "./type-safe-api-project";
export * from "./model/type-safe-api-model-project";
export * from "./model/smithy/smithy-definition";
export * from "./model/openapi/open-api-definition";
export * from "./type-safe-websocket-api-project";
export * from "./model/type-safe-websocket-api-model-project";
export * from "./model/smithy/smithy-async-definition";
export * from "./model/openapi/open-api-async-definition";
export * from "./model/smithy/types";
export * from "./model/openapi/open-api-model-project";
export * from "./model/openapi/open-api-async-model-project";
export * from "./model/smithy/smithy-async-model-project";
export * from "./model/smithy/smithy-model-project";
export * from "./languages";
export * from "./types";
export * from "./typescript-project-options";
export * from "./python-project-options";
export * from "./java-project-options";
