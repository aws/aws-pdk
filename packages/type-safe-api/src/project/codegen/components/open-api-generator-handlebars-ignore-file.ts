/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import { IgnoreFile, Project } from "projen";
import { OpenApiGeneratorIgnoreFile } from "./open-api-generator-ignore-file";

/**
 * Represents an .openapi-generator-ignore-handlebars file.
 * This is used for an optional second code generation pass to generate files using the handlebars templating engine
 * @see https://github.com/OpenAPITools/openapi-generator/blob/master/docs/customization.md#ignore-file-format
 */
export class OpenApiGeneratorHandlebarsIgnoreFile extends IgnoreFile {
  constructor(project: Project) {
    super(project, ".openapi-generator-ignore-handlebars");

    // Ignore all files by default to allow opting in to generate with specific handlebars templates
    this.addPatterns(...OpenApiGeneratorIgnoreFile.ALL_FILES_PATTERNS);
  }
}
