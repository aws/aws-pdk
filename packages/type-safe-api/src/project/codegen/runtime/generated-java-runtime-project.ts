/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as path from "path";
import { JavaProject } from "projen/lib/java";
import { Language } from "../../languages";
import { GeneratedJavaProjectOptions } from "../../types";
import { OpenApiGeneratorIgnoreFile } from "../components/open-api-generator-ignore-file";
import { OpenApiToolsJsonFile } from "../components/open-api-tools-json-file";
import {
  buildCleanOpenApiGeneratedCodeCommand,
  buildInvokeOpenApiGeneratorCommand,
} from "../components/utils";

/**
 * Configuration for the generated java runtime project
 */
export interface GeneratedJavaTypesProjectOptions
  extends GeneratedJavaProjectOptions {
  /**
   * The path to the OpenAPI specification, relative to this project's outdir
   */
  readonly specPath: string;
}

const DEPENDENCIES: string[] = [
  // Required for open api generated code
  "io.swagger/swagger-annotations@1.6.8",
  "com.google.code.findbugs/jsr305@3.0.2",
  "com.squareup.okhttp3/okhttp@4.10.0",
  "com.squareup.okhttp3/logging-interceptor@4.10.0",
  "com.google.code.gson/gson@2.9.1",
  "io.gsonfire/gson-fire@1.8.5",
  "org.apache.commons/commons-lang3@3.12.0",
  "jakarta.annotation/jakarta.annotation-api@1.3.5",
  "org.openapitools/jackson-databind-nullable@0.2.4",
  "javax.ws.rs/jsr311-api@1.1.1",
  "javax.ws.rs/javax.ws.rs-api@2.1.1",
  // For handler wrappers
  "com.amazonaws/aws-lambda-java-core@1.2.1",
  "com.amazonaws/aws-lambda-java-events@3.11.0",
  // Lombok is used to add the builder pattern to models for neater construction
  "org.projectlombok/lombok@1.18.24",
];

const TEST_DEPENDENCIES: string[] = [
  "org.junit.jupiter/junit-jupiter-api@5.9.1",
  "org.mockito/mockito-core@3.12.4",
];

/**
 * Java project containing types generated using OpenAPI Generator CLI
 */
export class GeneratedJavaRuntimeProject extends JavaProject {
  /**
   * Patterns that are excluded from code generation
   */
  public static openApiIgnorePatterns: string[] = [
    "pom.xml",
    "gradle",
    "gradle/**/*",
    "gradlew",
    "gradle.properties",
    "gradlew.bat",
    "build.gradle",
    "settings.gradle",
    "build.sbt",
    ".travis.yml",
    "git_push.sh",
    "src/test",
    "src/test/**/*",
    "src/main/AndroidManifest.xml",
  ];

  /**
   * The package name, for use in imports
   */
  public readonly packageName: string;

  /**
   * Path to the openapi specification
   * @private
   */
  private readonly specPath: string;

  constructor(options: GeneratedJavaTypesProjectOptions) {
    super({
      sample: false,
      junit: false,
      ...options,
    });
    this.specPath = options.specPath;

    // Ignore files that we will control via projen
    const ignoreFile = new OpenApiGeneratorIgnoreFile(this);
    ignoreFile.addPatterns(
      ...GeneratedJavaRuntimeProject.openApiIgnorePatterns
    );

    // Add OpenAPI Generator cli configuration
    OpenApiToolsJsonFile.ensure(this).addOpenApiGeneratorCliConfig(
      options.openApiGeneratorCliConfig
    );

    // Add dependencies
    DEPENDENCIES.forEach((dep) => this.addDependency(dep));
    TEST_DEPENDENCIES.forEach((dep) => this.addTestDependency(dep));

    this.packageName = `${this.pom.groupId}.${this.name}.runtime`;

    // Generate the java code
    const generateCodeCommand = this.buildGenerateCommand();
    const cleanCommand = buildCleanOpenApiGeneratedCodeCommand(this.outdir);

    const generateTask = this.addTask("generate");
    generateTask.exec(cleanCommand.command, {
      cwd: path.relative(this.outdir, cleanCommand.workingDir),
    });
    generateTask.exec(generateCodeCommand.command, {
      cwd: path.relative(this.outdir, generateCodeCommand.workingDir),
    });

    this.preCompileTask.spawn(generateTask);

    // Ignore all the generated code
    this.gitignore.addPatterns(
      "src",
      "docs",
      "api",
      "README.md",
      ".openapi-generator"
    );
  }

  public buildGenerateCommand = () => {
    return buildInvokeOpenApiGeneratorCommand({
      generator: "java",
      specPath: this.specPath,
      outputPath: this.outdir,
      generatorDirectory: Language.JAVA,
      additionalProperties: {
        useSingleRequestParameter: "true",
        groupId: this.pom.groupId,
        artifactId: this.pom.artifactId,
        artifactVersion: this.pom.version,
        invokerPackage: this.packageName,
        apiPackage: `${this.packageName}.api`,
        modelPackage: `${this.packageName}.model`,
        hideGenerationTimestamp: "true",
        additionalModelTypeAnnotations: [
          "@lombok.AllArgsConstructor",
          // Regular lombok builder is not used since an abstract base schema class is also annotated
          "@lombok.experimental.SuperBuilder",
        ].join("\\ "),
      },
      srcDir: path.join("src", "main", "java", ...this.packageName.split(".")),
      normalizers: {
        KEEP_ONLY_FIRST_TAG_IN_OPERATION: true,
      },
      // Do not generate map/list types. Generator will use built in HashMap, ArrayList instead
      generateAliasAsModel: false,
    });
  };
}
