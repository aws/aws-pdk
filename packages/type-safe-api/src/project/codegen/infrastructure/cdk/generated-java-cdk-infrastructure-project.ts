/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as path from "path";
import { DependencyType } from "projen";
import { JavaProject } from "projen/lib/java";
import {
  CodeGenerationSourceOptions,
  GeneratedJavaInfrastructureOptions,
} from "../../../types";
import { OpenApiGeneratorIgnoreFile } from "../../components/open-api-generator-ignore-file";
import { OpenApiToolsJsonFile } from "../../components/open-api-tools-json-file";
import {
  buildCleanOpenApiGeneratedCodeCommand,
  buildInvokeMockDataGeneratorCommand,
  buildInvokeOpenApiGeneratorCommandArgs,
  buildTypeSafeApiExecCommand,
  OtherGenerators,
  TypeSafeApiScript,
} from "../../components/utils";
import { GeneratedJavaRuntimeProject } from "../../runtime/generated-java-runtime-project";

export interface GeneratedJavaCdkInfrastructureProjectOptions
  extends GeneratedJavaInfrastructureOptions,
    CodeGenerationSourceOptions {
  /**
   * The generated java types
   */
  readonly generatedJavaTypes: GeneratedJavaRuntimeProject;
}

export class GeneratedJavaCdkInfrastructureProject extends JavaProject {
  /**
   * Options configured for the project
   * @private
   */
  private readonly options: GeneratedJavaCdkInfrastructureProjectOptions;

  /**
   * Source directory
   * @private
   */
  private readonly srcDir: string;

  /**
   * Java package name
   * @private
   */
  private readonly packageName: string;

  constructor(options: GeneratedJavaCdkInfrastructureProjectOptions) {
    super({
      ...options,
      sample: false,
      junit: false,
    });
    this.options = options;
    this.packageName = `${this.pom.groupId}.${this.name}.infra`;
    this.srcDir = path.join(
      "src",
      "main",
      "java",
      ...this.packageName.split(".")
    );

    [
      "software.aws.awsprototypingsdk/type-safe-api@^0",
      "software.constructs/constructs@^10",
      "software.amazon.awscdk/aws-cdk-lib@^2",
      "io.github.cdklabs/cdknag@^2",
      "org.projectlombok/lombok@^1",
      "com.fasterxml.jackson.core/jackson-databind@^2",
      "io.github.cdklabs/projen@^0",
      `${options.generatedJavaTypes.pom.groupId}/${options.generatedJavaTypes.pom.artifactId}@${options.generatedJavaTypes.pom.version}`,
    ]
      .filter(
        (dep) =>
          !this.deps.tryGetDependency(dep.split("@")[0], DependencyType.RUNTIME)
      )
      .forEach((dep) => this.addDependency(dep));

    // Remove the projen test dependency since otherwise it takes precedence, causing projen to be unavailable at synth time
    this.deps.removeDependency("io.github.cdklabs/projen", DependencyType.TEST);

    // Add a dependency on the generated java types repository
    this.pom.addRepository({
      url: `file://${path.relative(
        this.outdir,
        options.generatedJavaTypes.outdir
      )}/dist/java`,
      id: `${options.generatedJavaTypes.pom.groupId}-${options.generatedJavaTypes.pom.artifactId}-repo`,
    });

    // Ignore everything but the target files
    const openapiGeneratorIgnore = new OpenApiGeneratorIgnoreFile(this);
    openapiGeneratorIgnore.addPatterns(
      "/*",
      "**/*",
      "*",
      `!${this.srcDir}/Api.java`,
      `!${this.srcDir}/ApiProps.java`,
      `!${this.srcDir}/MockIntegrations.java`
    );

    // Add OpenAPI Generator cli configuration
    OpenApiToolsJsonFile.ensure(this).addOpenApiGeneratorCliConfig(
      options.openApiGeneratorCliConfig
    );

    const generateTask = this.addTask("generate");
    generateTask.exec(buildCleanOpenApiGeneratedCodeCommand());
    generateTask.exec(
      buildTypeSafeApiExecCommand(
        TypeSafeApiScript.GENERATE,
        this.buildGenerateCommandArgs()
      )
    );
    // Copy the parsed spec into the resources directory so that it's included in the jar
    generateTask.exec("mkdir -p src/main/resources");
    generateTask.exec(
      `cp -f ${this.options.specPath} src/main/resources/.api.json`
    );
    if (!this.options.mockDataOptions?.disable) {
      generateTask.exec(this.buildGenerateMockDataCommand());
    }

    this.preCompileTask.spawn(generateTask);

    // Ignore the generated code
    this.gitignore.addPatterns("src", ".openapi-generator");
  }

  public buildGenerateCommandArgs = () => {
    return buildInvokeOpenApiGeneratorCommandArgs({
      generator: "java",
      specPath: this.options.specPath,
      smithyJsonPath: this.options.smithyJsonModelPath,
      generatorDirectory: OtherGenerators.JAVA_CDK_INFRASTRUCTURE,
      srcDir: this.srcDir,
      normalizers: {
        KEEP_ONLY_FIRST_TAG_IN_OPERATION: true,
      },
      extraVendorExtensions: {
        "x-infrastructure-package": this.packageName,
        "x-runtime-package": this.options.generatedJavaTypes.packageName,
        // Enable mock integration generation by default
        "x-enable-mock-integrations": !this.options.mockDataOptions?.disable,
      },
      // Do not generate map/list types. Generator will use built in HashMap, ArrayList instead
      generateAliasAsModel: false,
    });
  };

  public buildGenerateMockDataCommand = (): string => {
    return buildInvokeMockDataGeneratorCommand({
      specPath: this.options.specPath,
      // Write the mocks to the resources directory
      outputSubDir: "src/main/resources",
      ...this.options.mockDataOptions,
    });
  };
}
