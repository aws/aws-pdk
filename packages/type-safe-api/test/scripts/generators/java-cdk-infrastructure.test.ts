/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as path from "path";
import { exec } from "projen/lib/util";
import { OpenApiToolsJsonFile } from "../../../src/project/codegen/components/open-api-tools-json-file";
import { GeneratedJavaCdkInfrastructureProject } from "../../../src/project/codegen/infrastructure/cdk/generated-java-cdk-infrastructure-project";
import { GeneratedJavaRuntimeProject } from "../../../src/project/codegen/runtime/generated-java-runtime-project";
import { withTmpDirSnapshot } from "../../project/snapshot-utils";

describe("Java Infrastructure Code Generation Script Unit Tests", () => {
  it.each(["single.yaml"])("Generates With %s", (spec) => {
    const specPath = path.resolve(__dirname, `../../resources/specs/${spec}`);

    const snapshot = withTmpDirSnapshot(
      path.resolve(__dirname),
      (outdir) => {
        const clientOutdir = path.join(outdir, "client");
        const client = new GeneratedJavaRuntimeProject({
          name: "test-client",
          artifactId: "com.aws.pdk.test.client",
          groupId: "test",
          version: "1.0.0",
          outdir: clientOutdir,
          specPath: path.relative(clientOutdir, specPath),
        });
        const infraOutdir = path.join(outdir, "infra");
        const project = new GeneratedJavaCdkInfrastructureProject({
          name: "test-infra",
          artifactId: "com.aws.pdk.test.infra",
          groupId: "test",
          version: "1.0.0",
          outdir: infraOutdir,
          specPath: path.relative(infraOutdir, specPath),
          generatedJavaTypes: client,
        });
        // Synth the openapitools.json since it's used by the generate command
        OpenApiToolsJsonFile.of(project)!.synthesize();
        const command = project.buildGenerateCommand();
        exec(command.command, {
          cwd: command.workingDir,
        });
      },
      {
        excludeGlobs: GeneratedJavaRuntimeProject.openApiIgnorePatterns,
        parseJson: false,
      }
    );

    expect(
      snapshot["infra/src/main/java/test/test-infra/infra/Api.java"]
    ).toMatchSnapshot();
    expect(
      snapshot["infra/src/main/java/test/test-infra/infra/ApiProps.java"]
    ).toMatchSnapshot();
    expect(
      snapshot[
        "infra/src/main/java/test/test-infra/infra/MockIntegrations.java"
      ]
    ).toMatchSnapshot();
  });
});
