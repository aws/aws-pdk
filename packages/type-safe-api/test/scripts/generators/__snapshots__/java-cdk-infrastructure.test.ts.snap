// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Java Infrastructure Code Generation Script Unit Tests Generates With single.yaml 1`] = `
"package test.test-infra.infra;

import com.fasterxml.jackson.databind.ObjectMapper;
import software.aws.awsprototypingsdk.typesafeapi.OperationDetails;
import software.aws.awsprototypingsdk.typesafeapi.TypeSafeRestApi;
import software.aws.awsprototypingsdk.typesafeapi.TypeSafeRestApiProps;
import software.constructs.Construct;
import test.test-client.runtime.api.OperationLookup;

import java.io.IOException;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Type-safe construct for the API Gateway resources defined by your model.
 * This construct is generated and should not be modified.
 */
public class Api extends TypeSafeRestApi {
    private static class SpecDetails {
        static String specPath;

        static {
            try {
                // The parsed spec is included as a packaged resource
                URL specUrl = SpecDetails.class.getClassLoader().getResource(".api.json");

                // We'll write the parsed spec to a temporary file outside of the jar to ensure CDK can package it as an asset
                Path parsedSpecPath = Files.createTempFile("type-safe-api", ".json");
                specPath = parsedSpecPath.toString();

                ObjectMapper json = new ObjectMapper();
                Object spec = json.readValue(specUrl, Object.class);
                json.writeValue(parsedSpecPath.toFile(), spec);
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
    }

    public Api(Construct scope, String id, ApiProps props) {
        super(scope, id, TypeSafeRestApiProps.builder()
                .defaultAuthorizer(props.getDefaultAuthorizer())
                .corsOptions(props.getCorsOptions())
                .apiKeyOptions(props.getApiKeyOptions())
                .operationLookup(OperationLookup.getOperationLookup()
                        .entrySet()
                        .stream()
                        .collect(Collectors.toMap(Map.Entry::getKey, e -> OperationDetails.builder()
                                .method(e.getValue().getMethod())
                                .path(e.getValue().getPath())
                                .contentTypes(e.getValue().getContentTypes())
                                .build())))
                .specPath(SpecDetails.specPath)
                .integrations(props.getIntegrations().asMap())
                // Rest API Base Props
                .cloudWatchRole(props.getCloudWatchRole())
                .deploy(props.getDeploy())
                .deployOptions(props.getDeployOptions())
                .description(props.getDescription())
                .disableExecuteApiEndpoint(props.getDisableExecuteApiEndpoint())
                .domainName(props.getDomainName())
                .endpointExportName(props.getEndpointExportName())
                .endpointTypes(props.getEndpointTypes())
                .failOnWarnings(props.getFailOnWarnings())
                .parameters(props.getParameters())
                .policy(props.getPolicy())
                .restApiName(props.getRestApiName())
                .retainDeployments(props.getRetainDeployments())
                .build());
    }
}
"
`;

exports[`Java Infrastructure Code Generation Script Unit Tests Generates With single.yaml 2`] = `
"package test.test-infra.infra;

import software.amazon.awscdk.services.apigateway.CorsOptions;
import software.amazon.awscdk.services.apigateway.StageOptions;
import software.amazon.awscdk.services.apigateway.RestApiBaseProps;
import software.amazon.awscdk.services.apigateway.DomainNameOptions;
import software.amazon.awscdk.services.apigateway.EndpointType;
import software.amazon.awscdk.services.iam.PolicyDocument;
import software.aws.awsprototypingsdk.typesafeapi.ApiKeyOptions;
import software.aws.awsprototypingsdk.typesafeapi.Authorizer;
import software.aws.awsprototypingsdk.typesafeapi.TypeSafeApiIntegration;

import test.test-client.runtime.api.OperationConfig;

import java.util.List;
import java.util.Map;

/**
 * Properties for the Api construct
 */
@lombok.Builder @lombok.Getter
public class ApiProps implements RestApiBaseProps {
    public OperationConfig<TypeSafeApiIntegration> integrations;
    public Authorizer defaultAuthorizer;
    public CorsOptions corsOptions;
    public ApiKeyOptions apiKeyOptions;

    // Rest API Props
    public Boolean cloudWatchRole;
    public Boolean deploy;
    public StageOptions deployOptions;
    public String description;
    public Boolean disableExecuteApiEndpoint;
    public DomainNameOptions domainName;
    public String endpointExportName;
    public List<EndpointType> endpointTypes;
    public Boolean failOnWarnings;
    public Map<String, String> parameters;
    public PolicyDocument policy;
    public String restApiName;
    public Boolean retainDeployments;
}
"
`;

exports[`Java Infrastructure Code Generation Script Unit Tests Generates With single.yaml 3`] = `
"package test.test-infra.infra;

import test.test-client.runtime.JSON;
import test.test-client.runtime.model.*;
import software.aws.awsprototypingsdk.typesafeapi.Integrations;
import software.aws.awsprototypingsdk.typesafeapi.MockIntegration;
import software.aws.awsprototypingsdk.typesafeapi.MockIntegrationResponse;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.file.Paths;
import java.util.stream.Collectors;


/**
 * Type-safe mock integrations for API operations
 */
public class MockIntegrations {
    static {
        // Instantiate json instance so that any .toJson() methods can be used
        new JSON();
    }

    /**
     * Read a mock data file for the given operation
     */
    private static String readMockDataFile(final String method, final String path, final int statusCode) {
        try {
            try (InputStream stream = MockIntegrations.class.getClassLoader()
                    .getResourceAsStream(Paths.get("mocks", method + path.replaceAll("/", "-") + String.format("-%d.json", statusCode)).toString())) {
                try (InputStreamReader inputStreamReader = new InputStreamReader(stream)) {
                    return new BufferedReader(inputStreamReader).lines().collect(Collectors.joining(System.lineSeparator()));
                }
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Mock integration to return a 200 response from the anyRequestResponse operation
     */
    public static MockIntegration anyRequestResponse200(final String body) {
        return Integrations.mock(MockIntegrationResponse.builder()
                .statusCode(200)
                .body(body)
                .build());
    }

    /**
     * Mock integration to return a 204 response from the empty operation
     */
    public static MockIntegration empty204() {
        return Integrations.mock(MockIntegrationResponse.builder()
                .statusCode(204)
                .build());
    }

    /**
     * Mock integration to return a 200 response from the mapResponse operation
     */
    public static MockIntegration mapResponse200(final MapResponse200Response body) {
        return Integrations.mock(MockIntegrationResponse.builder()
                .statusCode(200)
                .body(body.toJson())
                .build());
    }

    /**
     * Mock integration to return a 200 response from the mapResponse operation, with generated mock data
     */
    public static MockIntegration mapResponse200() {
        return Integrations.mock(MockIntegrationResponse.builder()
                .statusCode(200)
                .body(MockIntegrations.readMockDataFile("get", "/map-response", 200))
                .build());
    }

    /**
     * Mock integration to return a 200 response from the mediaTypes operation
     */
    public static MockIntegration mediaTypes200(final String body) {
        return Integrations.mock(MockIntegrationResponse.builder()
                .statusCode(200)
                .body(body)
                .build());
    }

    /**
     * Mock integration to return a 200 response from the multipleContentTypes operation
     */
    public static MockIntegration multipleContentTypes200(final String body) {
        return Integrations.mock(MockIntegrationResponse.builder()
                .statusCode(200)
                .body(body)
                .build());
    }

    /**
     * Mock integration to return a 200 response from the operationOne operation
     */
    public static MockIntegration operationOne200(final TestResponse body) {
        return Integrations.mock(MockIntegrationResponse.builder()
                .statusCode(200)
                .body(body.toJson())
                .build());
    }

    /**
     * Mock integration to return a 200 response from the operationOne operation, with generated mock data
     */
    public static MockIntegration operationOne200() {
        return Integrations.mock(MockIntegrationResponse.builder()
                .statusCode(200)
                .body(MockIntegrations.readMockDataFile("post", "/path/{pathParam}", 200))
                .build());
    }

    /**
     * Mock integration to return a 400 response from the operationOne operation
     */
    public static MockIntegration operationOne400(final ApiError body) {
        return Integrations.mock(MockIntegrationResponse.builder()
                .statusCode(400)
                .body(body.toJson())
                .build());
    }

    /**
     * Mock integration to return a 400 response from the operationOne operation, with generated mock data
     */
    public static MockIntegration operationOne400() {
        return Integrations.mock(MockIntegrationResponse.builder()
                .statusCode(400)
                .body(MockIntegrations.readMockDataFile("post", "/path/{pathParam}", 400))
                .build());
    }

    /**
     * Mock integration to return a 200 response from the withoutOperationIdDelete operation
     */
    public static MockIntegration withoutOperationIdDelete200(final TestResponse body) {
        return Integrations.mock(MockIntegrationResponse.builder()
                .statusCode(200)
                .body(body.toJson())
                .build());
    }

    /**
     * Mock integration to return a 200 response from the withoutOperationIdDelete operation, with generated mock data
     */
    public static MockIntegration withoutOperationIdDelete200() {
        return Integrations.mock(MockIntegrationResponse.builder()
                .statusCode(200)
                .body(MockIntegrations.readMockDataFile("delete", "/without-operation-id", 200))
                .build());
    }

}
"
`;
